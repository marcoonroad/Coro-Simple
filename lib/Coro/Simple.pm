use v6;

unit module Coro::Simple;

sub coro (&block) is export {
    return sub (*@params) {
        # we explore the lazy evaluation property
        # to evaluate the computation on demand
        my @yields := (gather { block |@params }).list;
        my $index   = 0;
        # generator
        return sub ( ) {
            my $result;
            if defined @yields[ $index++ ] {
                $result = @yields[ $index - 1 ];
            }
            # otherwise, if the coroutine is dead
            else {
        	$index -= 1; # disallows further lookups
                $result = False;
            }
            return $result;
        }
    }
}

sub yield ($value) is export { take $value }

sub suspend( ) is export { take True }

# purpose:
# to check if generated value was returned from yield triggering (rather end of computation)
sub ensure (&block, $value) is export {
    if ($value ~~ Bool) && (!$value) {
        return block;
    }
    return $value;
}

# deprecated function
sub assert (&block, $value) is export {
    warn "[ Deprecated function ]\n" ~
        "This function will be removed in a future version. Please use 'ensure' instead.";
    return ensure &block, $value;
}

sub from (&generator) is export {
    my $temp = generator; # obtains the first value

    return gather {
        while ($temp !~~ Bool) || (?$temp) {
            take $temp;
            $temp = generator;
        }
    };
}

# end of module
