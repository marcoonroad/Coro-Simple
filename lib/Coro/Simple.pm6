use v6;

unit module Coro::Simple;

sub coro (&block) is export {
  my $coroutine-thread = $*THREAD;

  return sub (*@params) {
    # we explore the lazy evaluation property
    # to evaluate the computation on demand
    my @yields := (gather { block |@params }).list;
    my $index   = 0;
    # generator
    return sub ( ) {
      unless $*THREAD.id == $coroutine-thread.id {
        die "Yield called outside the owner coroutine!";
      }

      my $result;
      if defined @yields[ $index++ ] {
        $result = @yields[ $index - 1 ];
      }
      # otherwise, if the coroutine is dead
      else {
	      $index -= 1; # disallows further lookups
        $result = False;
      }
      return $result;
    }
  }
}

sub yield ($value) is export { take $value }

sub suspend( ) is export { take True }

# purpose:
# to check if generated value was returned from yield triggering (rather end of computation)
sub ensure (&block, $value) is export {
  if ($value ~~ Bool) && (!$value) {
    return block;
  }
  return $value;
}

sub from (&generator) is export {
  my $temp = generator; # obtains the first value

  return gather {
    while ($temp !~~ Bool) || (?$temp) {
      take $temp;
      $temp = generator;
    }
  };
}

# end of module
