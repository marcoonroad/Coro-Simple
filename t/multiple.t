#!/usr/bin/env perl6

use v6;
use lib 'lib';
# a test that yields multiple values

use Test;
use Coro::Simple;

plan 8;

my $expected = slurp './t/expected/multiple.out';
my $handler = open './t/expected/multiple.out', :w;

my &transform = coro sub (&fn, *@xs) {
  for @xs -> $x, $y, $z {
    fn $x;
    fn $y;
    fn $z;
  }
}

my &get-next = transform -> $x {
  yield [ $x, $x + 1, $x ** 2 ] # will yields an anonymous list
}, (45 ... 15);

my $items;

for ^7 {
  $items = get-next;
  ok((defined $items), "generator result is non-null");
  $handler.say: "$items";
}

$handler.close;
my $current = slurp './t/expected/multiple.out';
is $expected, $current, "generated logs must match";

# end of test
