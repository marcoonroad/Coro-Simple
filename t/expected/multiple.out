45 46 2025
44 45 1936
43 44 1849
42 43 1764
41 42 1681
40 41 1600
39 40 1521
